packages/server/lib/cloud/environment.ts

```diff
@@ -0,0 +1,143 @@
+
+        if (cypressEnvUrl) {
+          acc.set(cypressEnvUrl[1], cypressEnvUrl[2])
+        }
+
+        return acc
+      }, new Map())
+
+      const foundPid = processBranch.find((pid) => pidEnvUrlMapping.get(pid))
+
+      if (foundPid) {
+        envUrl = pidEnvUrlMapping.get(foundPid)
+      }
+    } catch (err) {
+      error = err
+    }
+  }
+
+  return {
+    envUrl,
+    error,
+  }
+}
+
+const getEnvInformationForProjectRoot = async (projectRoot: string, pid: string) => {
+  let dependencies = {}
+  let errors: { dependency?: string, name: string, message: string, stack: string }[] = []
+  let envDependenciesVar = process.env.CYPRESS_ENV_DEPENDENCIES
+  let envUrl = process.env.CYPRESS_API_URL
+  let processTreePromise: Promise<GetCypressEnvUrlFromProcessBranch> = !envUrl ? getCypressEnvUrlFromProcessBranch(pid) : Promise.resolve({})
+
+  if (envDependenciesVar) {
+    const envDependenciesInformation = JSON.parse(base64Url.decode(envDependenciesVar)) as Record<string, { processTreeRequirement: 'presence required' | 'absence required' | 'irrelevant' }>
+
+    const packageToJsonMapping: Record<string, string> = {}
+
+    Object.entries(envDependenciesInformation).map(([dependency, { processTreeRequirement }]) => {
+      try {
+        const packageJsonPath = resolvePackagePath(dependency, projectRoot)
+
+        if (packageJsonPath) {
+          packageToJsonMapping[dependency] = packageJsonPath
+        }
+      } catch (error) {
+        errors.push({
+          dependency,
+          name: error.name,
+          message: error.message,
+          stack: error.stack,
+        })
+      }
+      if (processTreeRequirement === 'presence required' && !packageToJsonMapping[dependency]) {
+        processTreePromise = Promise.resolve({})
+      } else if (processTreeRequirement === 'absence required' && packageToJsonMapping[dependency]) {
+        processTreePromise = Promise.resolve({})
+      }
+    })
+
+    const [{ envUrl: processTreeEnvUrl, error: processTreeError }] = await Promise.all([
+      processTreePromise,
+      ...Object.entries(packageToJsonMapping).map(async ([dependency, packageJsonPath]) => {
+        try {
+          const packageVersion = (await fs.readJSON(packageJsonPath)).version
+
+          dependencies[dependency] = {
+            version: packageVersion,
+          }
+        } catch (error) {
+          errors.push({
+            dependency,
+            name: error.name,
+            message: error.message,
+            stack: error.stack,
+          })
+        }
+      }),
+    ])
+
+    if (processTreeEnvUrl || processTreeError) {
+      envUrl = processTreeEnvUrl
+      if (processTreeError) {
+        errors.push(processTreeError)
+      }
+    }
+  }
+
+  return {
+    ...(envUrl ? { envUrl } : {}),
+    ...(errors.length > 0 ? { errors } : {}),
+    ...(Object.keys(dependencies).length > 0 ? { dependencies } : {}),
```

Comment:
won't this omit empty props instead of always returning the prop we care about with a `null` value or empty object?

Reply:
Yeah I wasn't sure which we preferred. Always sending the objects even if they're empty or omit them if they are. Could save a smidgen on request size the second way, but I can go either way on it.